src/posts/gpu-pipeline/index.md

```diff
@@ -0,0 +1,223 @@
+   - Alternatively, wait for [sharding](https://zarr.readthedocs.io/en/stable/user-guide/performance.html#sharding) to be supported for GPU buffers in zarr-python?
+
+The plot below shows the read performance of the original dataset vs. the rechunked dataset (to optimal chunk size) vs. uncompressed zarr v3 dataset.
+
+TODO: ADD plot here.
+
+### Step 2: Reading with zarr-python v3 + kvikIO :open_book:
+
+The advent of [Zarr v3](https://zarr.dev/blog/zarr-python-3-release/) bought many improvements, including the ability to [read from Zarr stores to CuPy arrays (i.e. GPU memory)](https://github.com/zarr-developers/zarr-python/issues/2574).
+
+Specifically, you can use the [`zarr-python`](https://github.com/zarr-developers/zarr-python) driver to read data from zarr->CPU->GPU, or the [`kvikio`](https://github.com/rapidsai/kvikio) driver to read data from zarr->GPU directly!
+
+To benefit from these new features, we recommend installing:
+
+- [`zarr>=3.0.3`](https://github.com/zarr-developers/zarr-python/releases/tag/v3.0.3)
+- [`xarray>=2025.03.00`](https://github.com/pydata/xarray/releases/tag/v2025.03.0)
+- [`kvikio>=25.04.00`](https://github.com/rapidsai/kvikio/releases/tag/v25.04.00)
+
+Reading to GPU can be enabled by using the [`zarr.config.enable_gpu()`](https://zarr.readthedocs.io/en/v3.0.6/user-guide/gpu.html) setting like so:
+
+```python
+import cupy as cp
+import xarray as xr
+import zarr
+
+airt = xr.tutorial.open_dataset("air_temperature", engine="netcdf4")
+airt.to_zarr(store="/tmp/air-temp.zarr", mode="w", zarr_format=3, consolidated=False)
+
+with zarr.config.enable_gpu():
+    ds = xr.open_dataset("/tmp/air-temp.zarr", engine="zarr", consolidated=False)
+    assert isinstance(ds.air.data, cp.ndarray)
+```
+
+Note that using `engine="zarr"` like above would still result in data being loaded into CPU memory before it goes to GPU memory.
+If you prefer to bypass CPU memory, and have GPU Direct Storage (GDS) enabled, you can use the `kvikio` driver like so:
+
+```python
+import kvikio.zarr
+
+with zarr.config.enable_gpu():
+    store = kvikio.zarr.GDSStore(root="/tmp/air-temp.zarr")
+    ds = xr.open_dataset(filename_or_obj=store, engine="zarr")
+    assert isinstance(ds.air.data, cp.ndarray)
+```
+
+This will read the data directly from the Zarr store to GPU memory, bypassing CPU memory altogether. This is especially useful for large datasets, as it reduces the amount of data that needs to be transferred between CPU and GPU memory.
+
+[ TODO: add a figure showing this -- technically decompression is still done on CPU. ]
+
+(TODO ongoing work) Eventually with this [cupy-xarray Pull Request merged](https://github.com/xarray-contrib/cupy-xarray/pull/70) (based on earlier work at https://xarray.dev/blog/xarray-kvikio), this can be simplified to:
+
+```python
+import cupy_xarray
+
+ds = xr.open_dataset(filename_or_obj="/tmp/air-temp.zarr", engine="kvikio")
+assert isinstance(ds.air.data, cp.ndarray)
+```
```

Comment:
This could go in a future work section at the end

Reply:
Yeah, I'm not sure if this API is feasible or even desirable (have tried to implement this in https://github.com/xarray-contrib/cupy-xarray/pull/70, but no luck yet patching the buffer protocol). So ok to move this towards the end.